cmd/goal/interact.go

```diff
@@ -0,0 +1,542 @@
+		}
+
+		procFlags := cmd.Flags()
+		procFlags.SetInterspersed(true)
+		procArgs := make(map[string]interface{})
+		procAccounts := make(map[string]*string)
+		procApps := make(map[string]*uint64)
+		for _, arg := range proc.Args {
+			switch arg.Kind {
+			case "int", "integer":
+				procArgs[arg.Name] = procFlags.Uint64(arg.Name, 0, arg.Help)
+			default:
+				procArgs[arg.Name] = procFlags.String(arg.Name, "", arg.Help)
+			}
+		}
+		for _, account := range proc.Accounts {
+			procAccounts[account.Name] = procFlags.String(account.Name, "", account.Help)
+		}
+		for _, app := range proc.ForeignApps {
+			procApps[app.Name] = procFlags.Uint64(app.Name, 0, app.Help)
+		}
+		procFlags.Parse(os.Args[1:])
+
+		var inputs appCallInputs
+		for _, arg := range proc.Args {
+			var callArg appCallArg
+			callArg.Encoding = arg.Kind
+
+			v := procArgs[arg.Name]
+			s, ok := v.(*string)
+			if ok {
+				callArg.Value = *s
+			} else {
+				i, ok := v.(*uint64)
+				if ok {
+					// TODO this decodes and re-encodes redundantly
+					callArg.Value = strconv.FormatUint(*i, 10)
+				} else {
+					reportErrorf("Could not re-encode key %s", arg.Name)
+				}
+			}
+			inputs.Args = append(inputs.Args, callArg)
+		}
+		for _, account := range proc.Accounts {
+			var addr basics.Address
+			s := *procAccounts[account.Name]
+			if s == "" {
+				if account.Explicit {
+					reportErrorf("Required account %s not provided", account.Name)
+				}
+			} else {
+				var err error
+				addr, err = basics.UnmarshalChecksumAddress(s)
+				if err != nil {
+					reportErrorf("Could not unmarshal address %s", addr)
+				}
+			}
+			inputs.Accounts = append(inputs.Accounts, addr.String())
+		}
+		for _, app := range proc.ForeignApps {
+			inputs.ForeignApps = append(inputs.ForeignApps, *procApps[app.Name])
+		}
+
+		if proc.OnCompletion == "" {
+			proc.OnCompletion = "NoOp"
+		}
+		onCompletion := mustParseOnCompletion(proc.OnCompletion)
+		appAccounts := inputs.Accounts
+		foreignApps := inputs.ForeignApps
+
+		appArgs := make([][]byte, len(inputs.Args))
+		for i, arg := range inputs.Args {
+			rawValue, err := parseAppArg(arg)
+			if err != nil {
+				reportErrorf("Could not parse argument corresponding to '%s': %v", proc.Args[i].Name, err)
+			}
+			appArgs[i] = rawValue
+		}
+
+		if appIdx == 0 {
+			switch onCompletion {
+			case transactions.CloseOutOC, transactions.ClearStateOC:
+				reportWarnf("OnCompletion %s may be ill-formed when creating an application", onCompletion)
+			}
+		}
+
+		var approvalProg, clearProg []byte
+		var tx transactions.Transaction
+		var err error
+		if appIdx == 0 {
+			approvalProg, clearProg = mustParseProgArgs()
+			localSchema := header.Query.Local.ToStateSchema()
+			globalSchema := header.Query.Global.ToStateSchema()
+			tx, err = client.MakeUnsignedAppCreateTx(onCompletion, approvalProg, clearProg, globalSchema, localSchema, appArgs, appAccounts, foreignApps)
+		} else {
+			switch onCompletion {
```

Comment:
Could you maybe use `client.MakeUnsignedApplicationCallTx` to kill this switch?

Reply:
Did not know that was there; done.